dropshot/src/server.rs

```diff
@@ -173,6 +151,62 @@ impl HttpServer {
     }
 }
 
+/**
+ * A connection to a currently running Dropshot server.
+ */
+pub struct RunningHttpServer {
+    app_state: Arc<DropshotState>,
+    local_addr: SocketAddr,
+    join_handle: tokio::task::JoinHandle<Result<(), hyper::Error>>,
+    close_channel: tokio::sync::oneshot::Sender<()>,
+}
+
+/*
+ * TODO: This struct would be a great target for async drop, if that happens - it
+ * would be great to have a drop implementation which can invoke terminate if
+ * the caller has not done so explicitly.
+ */
+impl RunningHttpServer {
+    pub fn local_addr(&self) -> SocketAddr {
+        self.local_addr
+    }
+
+    pub fn app_private(&self) -> Arc<dyn Any + Send + Sync + 'static> {
+        Arc::clone(&self.app_state.private)
+    }
+
+    /**
+     * Signals the currently running server to terminate (if it hasn't
+     * stopped already) and waits for it to exit.
+     */
+    pub async fn terminate(self) -> Result<(), String> {
```

Comment:
Summarizing for myself here: before this change, there were two interfaces that were part of shutdown:

- close(), which initiates shutdown but does not wait for it
- wait_for_shutdown(), which waits for the shutdown to finish (but doesn't initiate it)

It took me a while reviewing this to remember the reason for decoupling these: there are consumers where it makes sense to `wait_for_shutdown()` without calling `close()` -- namely, a standard server program, including all the examples in this repo.  There are also potentially use cases where `close()` might be invoked a different context than the one that created and waited on the server.  I'm not sure if this second case worked before, and it's not important to support that now if we don't use it already.

In the new code, we have `terminate()` and `with_graceful_shutdown()`.  `terminate()` initiates shutdown and waits for it, which is fine.  But for the standard server program case that I mentioned above, we still need a way to wait-until-shutdown without actually starting a shutdown.  I gather that's why `with_graceful_shutdown()` exists, but this interface feels awkward to me.  If I remember correctly, @cbiffle and I both found `with_graceful_shutdown()` in hyper to be confusing, both in its name and behavior.  I was trying to create a more intuitive interface in Dropshot.  One thing that's awkward here: if you're looking at this API, it's not obvious why a caller wouldn't just wait for the future themselves and then call terminate() -- this function doesn't seem to add anything that the caller couldn't do themselves.

If we end up keeping `terminate()`, I'd like the name to be a bit more explicit that it waits for shutdown to complete.  What about `terminate_and_wait`?  (Or `close_and_wait`.  Why the change from `close` to `terminate`?)

Reply:
> Summarizing for myself here: before this change, there were two interfaces that were part of shutdown:
> 
> * close(), which initiates shutdown but does not wait for it
> * wait_for_shutdown(), which waits for the shutdown to finish (but doesn't initiate it)

That's accurate, this is the old version:

```rust
pub fn close(self) { ... }

pub async fn wait_for_shutdown<'_>(
    &'_ mut self,
    join_handle: JoinHandle<Result<(), Error>>
) -> Result<(), String> { ... }
```

My problem with the old implementation was the really loose coupling of the `join_handle` with the actual server.

`wait_for_shutdown` never acts on `self`, so it can be invoked any number of times with totally unrelated handles.

> It took me a while reviewing this to remember the reason for decoupling these: there are consumers where it makes sense to `wait_for_shutdown()` without calling `close()` -- namely, a standard server program, including all the examples in this repo. There are also potentially use cases where `close()` might be invoked a different context than the one that created and waited on the server. I'm not sure if this second case worked before, and it's not important to support that now if we don't use it already.
> 
> In the new code, we have `terminate()` and `with_graceful_shutdown()`. `terminate()` initiates shutdown and waits for it, which is fine. But for the standard server program case that I mentioned above, we still need a way to wait-until-shutdown without actually starting a shutdown. I gather that's why `with_graceful_shutdown()` exists, but this interface feels awkward to me. If I remember correctly, @cbiffle and I both found `with_graceful_shutdown()` in hyper to be confusing, both in its name and behavior. I was trying to create a more intuitive interface in Dropshot. 
> One thing that's awkward here: if you're looking at this API, it's not obvious why a caller wouldn't just wait for the future themselves and then call terminate() -- this function doesn't seem to add anything that the caller couldn't do themselves.

I think this is a good point, and I agree with you - I've updated the API to the following:

- HttpServer implements `Future`, which is the primary mechanism to await the completion of the server. This basically punts as much control of the server lifetime as possible to the client.
- I've kept the async, self-consuming `close` function (sticking with that name instead of `terminate` to minimize changes) for the same hard-stop behavior.


> If we end up keeping `terminate()`, I'd like the name to be a bit more explicit that it waits for shutdown to complete. What about `terminate_and_wait`? (Or `close_and_wait`. Why the change from `close` to `terminate`?)

I have no strong attachment to the name `terminate` - happy to update to `close`.

It does seem quirky to have an async method named "XXX_and_wait" though - this seems like it should be implicit within the future.

I read:

`do_foo_and_wait().await;`

and

`do_foo().await;`

as doing the same thing.

With the updated version, do you still feel strongly about this?

Reply:
I don't feel strongly about it.